### ISSUE TITLE
Contrastive Loss

### ISSUE THREAD

**Issue by nikitamehrotra12** [KevinMusgrave/pytorch-metric-learning#14]
Hi,

Could you please post an example of using contrastive loss without trainers and miners, it's quite different from the contrastive loss that uses Euclidean distance between pairs?  

And also the reference from where the definition of this loss has been taken.

Thanks

**Comment by KevinMusgrave** [comment]
Here's an example: 
```python
from pytorch_metric_learning import losses
loss_func = losses.ContrastiveLoss()

for i, (imgs, labels) in enumerate(dataloader):
    embeddings = your_model(imgs)
    loss = loss_func(embeddings, labels)
```

As for the paper, I think I put the wrong one in the README, so I'll fix that. Check out equation 4 in this paper: http://yann.lecun.com/exdb/publis/pdf/hadsell-chopra-lecun-06.pdf.

The only difference in my implementation is that I allow for a positive pair margin to be specified.

**Comment by nikitamehrotra12** [comment]
Thank you for your quick response and help :)

**Comment by nikitamehrotra12** [comment]
In the above example do embeddings corresponds to the list i/p embeddings like [x1,x2]?

**Comment by KevinMusgrave** [comment]
Can you clarify what you mean by "list i/p embeddings like [x1, x2]"?

**Comment by nikitamehrotra12** [comment]
The siamese network returns me an embedding vector for both inputs. So here I am not sure of passing those embeddings as input to the loss function. o I pass the euclidean or cosine distance between the two or the concatenated vector of the two embeddings or the list as above?

**Comment by KevinMusgrave** [comment]
With this library, you don't need to be concerned about packaging your embeddings as pairs/triplets, and you don't need to compute pairwise distances etc. 

In the example above, let's say ```imgs``` is of size (5,3,256,256), i.e. there are 5 RGB images of size 256x256. A convnet (with its final classification layer removed) will output ```embeddings``` of size (5, D), where D might be something like 512 or 1024. Now let's say that ```labels``` contains something like ```torch.LongTensor([0,1,1,0,1])```. Then when you do ```loss_func(embeddings, labels)```, the ContrastiveLoss will automatically form all possible positive and negative pairs (based on the given labels), compute their distances, and compute the loss based on those distances.

**Comment by nikitamehrotra12** [comment]
Ohh! my dataset is slightly different...I am working with source code on clone detection...So I have embeddings in pairs as input and its corresponding label

**Comment by KevinMusgrave** [comment]
What shape is ```embeddings``` and what shape is ```labels```?

**Comment by nikitamehrotra12** [comment]
embeddings for a single input is in shape 2 X 1 X 64 and label is 1 dimensional (0 or 1)

**Comment by KevinMusgrave** [comment]
@nikitamehrotra12 Ok, you can reshape your variables like this:
```python
# Let's say there are 10 pairs
# Then I assume embeddings has shape 2x10x64
# And labels will have shape 10
embeddings = embeddings.view(-1,64)
labels = labels.repeat(2)
loss = loss_func(embeddings, labels)
```
